samples/web/content/apprtc/join_page_test.py

```diff
@@ -0,0 +1,287 @@
+from google.appengine.api import memcache
+from google.appengine.ext import testbed
+
+class AppRtcPageHandlerTest(unittest.TestCase):
+  def setUp(self):
+    # First, create an instance of the Testbed class.
+    self.testbed = testbed.Testbed()
+
+    # Then activate the testbed, which prepares the service stubs for use.
+    self.testbed.activate()
+
+    # Create a consistency policy that will simulate the High Replication
+    # consistency model.
+    self.policy = datastore_stub_util.PseudoRandomHRConsistencyPolicy()
+    # Initialize the datastore stub with this policy.
+    self.testbed.init_datastore_v3_stub(consistency_policy=self.policy)
+    # Next, declare which service stubs you want to use.
+    self.testbed.init_memcache_stub()
+
+    self.test_app = webtest.TestApp(apprtc.app)
+
+  def tearDown(self):
+    self.testbed.deactivate()
+
+  def makeGetRequest(self, path):
+    # PhantomJS uses WebKit, so Safari is closest to the thruth.
+    return self.test_app.get(path, headers={'User-Agent': 'Safari'})
+
+  def makePostRequest(self, path, body=''):
+    return self.test_app.post(path, body, headers={'User-Agent': 'Safari'})
+    
+  def addTestData(self):
+    gcmrecord.GCMRecord.add_or_update('caller1', 'caller1gcm1', 'caller1code1')
+    gcmrecord.GCMRecord.add_or_update('caller1', 'caller1gcm2', 'caller1code2')
+    gcmrecord.GCMRecord.verify('caller1', 'caller1gcm1', 'caller1code1')
+    gcmrecord.GCMRecord.verify('caller1', 'caller1gcm2', 'caller1code2')
+    gcmrecord.GCMRecord.add_or_update('callee1', 'callee1gcm1', 'callee1code1')
+    gcmrecord.GCMRecord.add_or_update('callee1', 'callee1gcm2', 'callee1code2')
+    gcmrecord.GCMRecord.add_or_update('callee1', 'callee1gcm3', 'callee1code3')
+    gcmrecord.GCMRecord.verify('callee1', 'callee1gcm1', 'callee1code1')
+    gcmrecord.GCMRecord.verify('callee1', 'callee1gcm2', 'callee1code2')
+    gcmrecord.GCMRecord.verify('callee1', 'callee1gcm3', 'callee1code3')
+    gcmrecord.GCMRecord.add_or_update('caller2', 'caller2gcm1', 'caller2code1')
+    gcmrecord.GCMRecord.verify('caller2', 'caller2gcm1', 'caller2code1')
+    gcmrecord.GCMRecord.add_or_update('callee2', 'callee2gcm1', 'callee2code1')
+    gcmrecord.GCMRecord.verify('callee2', 'callee2gcm1', 'callee2code1')
+    # Unverified caller and callee.
+    gcmrecord.GCMRecord.add_or_update('caller3', 'caller3gcm1', 'caller3code1')
+    gcmrecord.GCMRecord.add_or_update('callee3', 'callee3gcm1', 'callee3code1')
+
+  def verifyResultCode(self, response, expectedCode):
+    self.assertEqual(expectedCode, json.loads(response.body)['result'])
+
+  def testJoinAsCallerInvalidCaller(self):
+    self.addTestData()
+    
+    room_id = 'callercallee'
+    # Caller that doesn't exist.
+    body = {
+      apprtc.PARAM_ACTION: 'call',
+      apprtc.PARAM_CALLER_GCM_ID: 'foo',
+      apprtc.PARAM_CALLEE_ID: 'callee1'
+    }
+    
+    response = self.makePostRequest('/join/' + room_id, json.dumps(body))
```

Comment:
This pattern (body = ...; makePostRequest...; expect XXX) is used over and over again. Put it in a helper method?